### PR TITLE
`Generator`: add support for sampling

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This change adds support for sampling to the generator. When sampling is used, we draw a random piece using the given softmax distribution. This adds more variety to the output than greedy/deterministic argmax generation.

To offer some initial control over the sampling, this change also adds top-k sampling and softmax temperature. These two techniques transform the logits before sampling. In order to allow such transforms, we introduce a new base class `LogitsTransform`, with derived `TopKTransform` and `TemperatureTransform` classes.

From the user's perspective, top-k and temperature are configured through the new `SampleGenerationConfig` class. For greedy generation, we add `GreedyGenerationConfig`.

## Checklist

- [ ] I confirm that I have the right to submit this contribution under the project's MIT license.
- [ ] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/generation/logits.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+
+class LogitsTransform(ABC):
+    """
+    A logits transform changes the logits of a softmax distribution in some
+    way. For instance, :class:`TemperatureTransform` changes the temperature
+    of the softmax distribution.
+    """
+
+    def __call__(self, logits: Tensor, inplace=False) -> Tensor:
```
Comment (shadeMe) [inline curated_transformers/generation/logits.py:14]:
```suggestion
    def __call__(self, logits: Tensor, inplace: bool = False) -> Tensor:
```

File: curated_transformers/generation/logits.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        Transform the given logits.
+
+        :param logits:
+            The logits array.
+            **Shape:** (..., n_class)
+        :param inplace:
+            Transform logits in-place.
+        :returns:
+            The transformed logits. This will be the same tensor object as
+            the ``logits`` argument when ``inplace`` is set.
```
Comment (shadeMe) [inline curated_transformers/generation/logits.py:25]: Are the double back-ticks intentional?

File: curated_transformers/generation/logits.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    def __init__(self, k: int):
+        """
+        Construct a top-k logits transform.
+
+        :param k:
+            The value of k in top-k. The transform is a noop for values
+            less than 1.
+        """
+        super().__init__()
+        self.k = k
```
Comment (shadeMe) [inline curated_transformers/generation/logits.py:71]: What speaks against making `k>=1` an requirement?

File: curated_transformers/generation/logits.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        p(y|x) = \frac{\mbox{exp}(\frac{z_i}{T})}{\sum_{y'=1}^N{\mbox{exp}(\frac{z_{y'}}{T})}}
+
+    For a temperature T:
+
+    - T = 1: the distribution is not changed.
+    - T < 1: the entropy of the distribution is decreased.
+    - T > 1: the entropy of the distribution is increased.
+    """
+
+    def __init__(self, temperature=1.0):
```
Comment (shadeMe) [inline curated_transformers/generation/logits.py:100]:
```suggestion
    def __init__(self, temperature: float = 1.0):
```

File: curated_transformers/tests/generation/test_dolly_v2.py [hunk 5]
```diff
@@ -26,7 +33,35 @@ def test_generate_deterministic():
...
+        "What is spaCy?",
+    ]
+    # Fix the seed so that we are always randomly sampling in the same way.
+    torch.manual_seed(0)
+    assert dolly_generator(prompts, config=SampleGeneratorConfig(top_k=10)) == [
+        "SpaCy (short for Spanish-Portuguese-Chinese artificial intelligence language model) is an open-source language model (LLM) that uses a novel combination of deep learning, big-data, and semantic parsing to provide the best accuracy available with no human involvement. It is one of the most effective open-source language models.\n\n",
+        "SpaCy is an open-source package for Python that provides language model (LM) and automatic speech recognition (ASR) functionality.\n\n",
+    ]
+
+    torch.manual_seed(0)
```
Comment (shadeMe) [inline curated_transformers/tests/generation/test_dolly_v2.py:56]: Do we need to set this multiple times?

File: curated_transformers/generation/logits.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/generation/logits.py:71]: You want some value of `k` where you can effectively disable it. It seems that other implementations use `k<1` to do this. You could set `k` to a large value, bet then you need to know the vocab size.

File: curated_transformers/generation/config.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    output.
+
+    :param temperature:
+        Softmax temperature. For a temperature T:
+
+        - T = 1: the distribution is not changed.
+        - T < 1: the entropy of the distribution is decreased.
+        - T > 1: the entropy of the distribution is increased.
+    :param top_k:
+        Sample from top-k highest-probability pieces.
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/config.py]: Fixed.

File: curated_transformers/tests/generation/test_dolly_v2.py (hunk 5 above)
Comment (danieldk) [reply curated_transformers/tests/generation/test_dolly_v2.py:56]: Removed.

**Review by shadeMe — APPROVED** [review]
